druid-shell/src/dialog.rs

```diff
@@ -32,25 +34,98 @@ pub enum FileDialogType {
 }
 
 /// Options for file dialogs.
+///
+/// File dialogs let the user choose a specific path to open or save.
+///
+/// By default the file dialogs operate in *files mode* where the user can only choose files.
+/// Importantly these are files from the user's perspective, but technically the returned path
+/// will be a directory when the user chooses a package. Read more about [packages] below.
+/// It's also possible for users to manually specify a path which they might otherwise not be able
+/// to choose. Thus it is important to verify that all the returned paths match your expectations.
+///
+/// The open dialog can also be switched to *directories mode* via [`select_directories`].
+///
+/// # Packages
+///
+/// On macOS directories with known extensions are considered to be packages.
+/// Furthermore the packages are divided into two groups based on their extension.
+/// First there are packages that have been defined at the OS level, and secondly there are
+/// packages that are defined at the file dialog level based on [`allowed_types`].
+/// A package behaves similarly to a regular file in many contexts, including the file dialogs.
+/// This package concept can be turned off in the file dialog via [`packages_as_directories`].
+///
+/// &#xFEFF; | Packages as files. File filters apply to packages. | Packages as directories.
+/// -------- | -------------------------------------------------- | ------------------------
+/// Open directory | Not selectable. Not traversable. | Selectable. Traversable.
+/// Open file | Selectable. Not traversable. | Not selectable. Traversable.
+/// Save file | OS packages [clickable] but not traversable.<br/>Dialog packages traversable but not selectable. | Not selectable. Traversable.
+///
+/// Keep in mind that the file dialog may start inside any package if the user has traversed
+/// into one just recently. The user might also manually specify a path inside a package.
+///
+/// Generally this behavior should be kept, because it's least surprising to macOS users.
+/// However if your application requires selecting directories with extensions as directories
+/// or the user needs to be able to traverse into them to select a specific file,
+/// then you can change the default behavior via [`packages_as_directories`]
+/// to force macOS to behave like other platforms and not give special treatment to packages.
+///
+/// # macOS
```

Comment:
This differentiation between `Packages` and `macOS` kind of makes it look as if `Packages` would be relevant on other platforms as well.

Reply:
Good point, I moved it all under the `macOS` header.